File: src/core/backup/backup.ts
/**
 * 应用远端变更前的本地文件备份 + 备份保留策略（docs/m2-plan.md §2.3，P1-W2）。
 *
 * 冻结签名：
 *   BackupTarget / BackupResult
 *   backupFiles(paths, command, targets): BackupResult
 *   pruneBackups(paths, keep?): { removed: string[]; kept: string[] }
 *
 * 语义（计划原文）：
 *   - 目录布局：`<backupsDir>/<YYYYMMDD>/<HHmmss>-<command>/<label>`
 *     （`<label>` 形如 `pi/settings/settings.json`，保留 store 相对结构，便于人工核对与整体回填）。
 *   - 只备份**已存在**的源文件；源不存在的 target 记入 `skipped`（无需备份，调用方常直接跳过）。
 *   - `label` 必须是受控的相对路径（禁绝对路径 / 禁 `..` 段），否则 throw——备份目录由 label
 *     拼出绝对路径，label 一旦可逃逸就可能写到 `~` 或更远处（与 store.ts 的 assertSafeSegment 同因）。
 *     校验在**任何写操作之前**完成：非法 target 不会留下半套备份。
 *   - `pruneBackups`：按日期目录名（`YYYYMMDD`）降序保留前 `keep` 个，删除其余**整个日期目录**；
 *     `keep` 默认 7。返回的 `kept`/`removed` 是日期目录**名**（相对 `backupsDir`，非绝对路径）。
 *     backupsDir 不存在 → 空结果（幂等，不 throw）。非日期名的条目（例如人工遗留的临时目录）不参与保留策略。
 *
 * 备份不是同步内容的一部分（§1 决策 D5）：`backups/` 由 `~/.homer/.gitignore` 排除。
 */

import fs from 'node:fs';
import path from 'node:path';

import type { HomerPaths } from '../paths.js';

/** 一个待备份项：`sourceAbs` 为源文件绝对路径，`label` 为 store 相对路径形式的目标名。 */
export interface BackupTarget {
  sourceAbs: string;
  label: string; // 'pi/settings/settings.json'，禁 '..' / 绝对路径
}

/** `backedUp` / `skipped` 的元素均为对应 target 的 `label`。 */
export interface BackupResult {
  backupDir: string;
  backedUp: string[];
  skipped: string[]; // 源不存在（无需备份）
}

/** 日期目录名（`YYYYMMDD`），只有这种名字的目录参与保留策略。 */
const DATE_DIR_RE = /^\d{8}$/;

/** 默认保留最近 7 个日期目录（§2.0-1 `backup.keep` 缺省值）。 */
export const DEFAULT_BACKUP_KEEP = 7;

function two(n: number): string {
  return String(n).padStart(2, '0');
}

/** 本地时区的 `YYYYMMDD`（目录名与用户 `ls` 时看到的日期一致，故不用 UTC）。 */
function dateStamp(now: Date): string {
  return `${now.getFullYear()}${two(now.getMonth() + 1)}${two(now.getDate())}`;
}

/** 本地时区的 `HHmmss`。 */
function timeStamp(now: Date): string {
  return `${two(now.getHours())}${two(now.getMinutes())}${two(now.getSeconds())}`;
}

function assertSafeSegment(segment: string, what: string): void {
  if (segment.length === 0) throw new Error(`${what} 不能为空`);
  if (path.isAbsolute(segment) || segment === '..' || segment.startsWith(`..${path.sep}`)) {
    throw new Error(`${what} 非法（不得为绝对路径或包含 '..'）: ${segment}`);
  }
  if (segment.split(/[\\/]/).includes('..')) {
    throw new Error(`${what} 非法（不得包含 '..' 段）: ${segment}`);
  }
}

/**
 * 备份 `targets` 中已存在的源文件。
 * `command` 是产生备份的命令名（push/pull/merge），进目录名 `<HHmmss>-<command>`。
 * 同秒内重复调用会落到同一目录（互相覆盖）——计划未要求唯一化，调用方按命令粒度使用即可。
 */
export function backupFiles(
  paths: HomerPaths,
  command: string,
  targets: readonly BackupTarget[],
): BackupResult {
  assertSafeSegment(command, 'command');
  for (const target of targets) {
    assertSafeSegment(target.label, 'label');
  }

  const now = new Date();
  const backupDir = path.join(paths.backupsDir, dateStamp(now), `${timeStamp(now)}-${command}`);

  // 目录总是创建：`BackupResult.backupDir` 因此恒为「可展示给用户的真实路径」，
  // 即使本次无文件可备份（全 skipped）也能据此汇报「备份位置」。
  fs.mkdirSync(backupDir, { recursive: true });

  const backedUp: string[] = [];
  const skipped: string[] = [];

  for (const target of targets) {
    if (!fs.existsSync(target.sourceAbs)) {
      skipped.push(target.label);
      continue;
    }
    const dest = path.join(backupDir, target.label);
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    // recursive 同时覆盖目录型源（分类目录整体备份）与单文件源。
    fs.cpSync(target.sourceAbs, dest, { recursive: true });
    backedUp.push(target.label);
  }

  return { backupDir, backedUp, skipped };
}

/**
 * 保留最近 `keep` 个日期目录，删除更旧的整个目录。
 * 返回的数组元素是日期目录名（`YYYYMMDD`，相对 `backupsDir`）。
 * `keep=0` → 删除全部日期目录。非法 `keep`（负数 / 非整数）→ throw（编程错误）。
 */
export function pruneBackups(
  paths: HomerPaths,
  keep: number = DEFAULT_BACKUP_KEEP,
): { removed: string[]; kept: string[] } {
  if (!Number.isInteger(keep) || keep < 0) {
    throw new Error(`pruneBackups: keep 必须是非负整数（当前: ${JSON.stringify(keep)}）`);
  }

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(paths.backupsDir, { withFileTypes: true });
  } catch {
    // backupsDir 不存在（从未备份过）：无事可做。
    return { removed: [], kept: [] };
  }

  const names = entries
    .filter((entry) => entry.isDirectory() && DATE_DIR_RE.test(entry.name))
    .map((entry) => entry.name)
    .sort((a, b) => (a < b ? 1 : a > b ? -1 : 0)); // 名降序 = 日期新→旧

  const kept = names.slice(0, keep);
  const removed = names.slice(keep);
  for (const name of removed) {
    fs.rmSync(path.join(paths.backupsDir, name), { recursive: true, force: true });
  }

  return { removed, kept };
}

File: internal/backup/backup.go
// Package backup implements the on-disk backup tree used before applying
// remote changes to an adapter. Backups are deliberately outside the store.
package backup

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/zzjcool/homer-cli/internal/core"
)

// BackupTarget maps an existing source path to a controlled, store-relative
// label in the backup directory.
type BackupTarget struct {
	SourceAbs string
	Label     string
}

// BackupResult reports the exact backup directory and target outcomes.
type BackupResult struct {
	BackupDir string
	BackedUp  []string
	Skipped   []string
}

// BackupFileModes optionally tightens permissions on the backup tree. A zero
// field means "do not change that kind of entry"; this makes the ordinary
// (unspecified) path preserve the platform's default mode and umask behavior.
type BackupFileModes struct {
	Dir  fs.FileMode
	File fs.FileMode
}

// mode is the package-local spelling used by the frozen opts.mode notation.
// Its fields mirror BackupFileModes; it is intentionally not exported because
// callers outside this package should use the explicit type above.
type mode struct {
	dir  fs.FileMode
	file fs.FileMode
}

// BackupFilesOptions is accepted as an optional argument to BackupFiles. Mode
// may be either BackupFileModes or *BackupFileModes; accepting both keeps call
// sites concise while retaining an explicit opt-in for secret backups.
type BackupFilesOptions struct {
	Mode any
	mode any // package-local compatibility spelling; exported callers use Mode
}

const DefaultBackupKeep = 7

// DEFAULT_BACKUP_KEEP preserves the migration-facing constant spelling.
const DEFAULT_BACKUP_KEEP = DefaultBackupKeep

// BackupMode and BackupOptions are concise aliases for callers that use the
// frozen opts.mode terminology.
type BackupMode = BackupFileModes
type BackupOptions = BackupFilesOptions

var dateDirectoryPattern = regexp.MustCompile(`^\d{8}$`)

func twoDigits(value int) string {
	if value < 10 {
		return "0" + fmt.Sprint(value)
	}
	return fmt.Sprint(value)
}

func dateStamp(now time.Time) string {
	return fmt.Sprintf("%04d%s%s", now.Year(), twoDigits(int(now.Month())), twoDigits(now.Day()))
}

func timeStamp(now time.Time) string {
	return twoDigits(now.Hour()) + twoDigits(now.Minute()) + twoDigits(now.Second())
}

func assertSafeSegment(value, label string) error {
	if value == "" {
		return fmt.Errorf("%s 不能为空", label)
	}
	if filepath.IsAbs(value) {
		return fmt.Errorf("%s 非法（不得为绝对路径或包含 '..'）: %s", label, value)
	}
	for _, part := range strings.FieldsFunc(value, func(r rune) bool { return r == '/' || r == '\\' }) {
		if part == ".." {
			return fmt.Errorf("%s 非法（不得包含 '..' 段）: %s", label, value)
		}
	}
	return nil
}

func modeFromOptions(options []any) (BackupFileModes, bool, error) {
	if len(options) == 0 || options[0] == nil {
		return BackupFileModes{}, false, nil
	}
	if len(options) > 1 {
		return BackupFileModes{}, false, fmt.Errorf("backupFiles: opts 只能传一个")
	}
	switch value := options[0].(type) {
	case BackupFilesOptions:
		if value.Mode != nil {
			return modeFromOptions([]any{value.Mode})
		}
		return modeFromOptions([]any{value.mode})
	case *BackupFilesOptions:
		if value == nil {
			return BackupFileModes{}, false, nil
		}
		if value.Mode != nil {
			return modeFromOptions([]any{value.Mode})
		}
		return modeFromOptions([]any{value.mode})
	case mode:
		return BackupFileModes{Dir: value.dir, File: value.file}, true, nil
	case *mode:
		if value == nil {
			return BackupFileModes{}, false, nil
		}
		return BackupFileModes{Dir: value.dir, File: value.file}, true, nil
	case BackupFileModes:
		return value, true, nil
	case *BackupFileModes:
		if value == nil {
			return BackupFileModes{}, false, nil
		}
		return *value, true, nil
	default:
		return BackupFileModes{}, false, fmt.Errorf("backupFiles: opts.mode 类型无效")
	}
}

func chmodIfPresent(name string, mode fs.FileMode) error {
	if mode == 0 {
		return nil
	}
	if err := os.Chmod(name, mode.Perm()); err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	return nil
}

func applyModes(root string, modes BackupFileModes) error {
	if err := chmodIfPresent(root, modes.Dir); err != nil {
		return err
	}
	return filepath.WalkDir(root, func(name string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if name == root {
			return nil
		}
		if entry.IsDir() {
			return chmodIfPresent(name, modes.Dir)
		}
		return chmodIfPresent(name, modes.File)
	})
}

func copyPath(source, destination string) error {
	info, err := os.Stat(source)
	if err != nil {
		return err
	}
	if info.IsDir() {
		if err := os.MkdirAll(destination, 0o777); err != nil {
			return err
		}
		return filepath.WalkDir(source, func(name string, entry fs.DirEntry, walkErr error) error {
			if walkErr != nil {
				return walkErr
			}
			relative, err := filepath.Rel(source, name)
			if err != nil {
				return err
			}
			if relative == "." {
				return nil
			}
			target := filepath.Join(destination, relative)
			if entry.IsDir() {
				return os.MkdirAll(target, 0o777)
			}
			if !entry.Type().IsRegular() {
				return fmt.Errorf("不支持备份非普通文件: %s", name)
			}
			return copyRegularFile(name, target)
		})
	}
	if !info.Mode().IsRegular() {
		return fmt.Errorf("不支持备份非普通文件: %s", source)
	}
	return copyRegularFile(source, destination)
}

func copyRegularFile(source, destination string) error {
	data, err := os.ReadFile(source)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(destination), 0o777); err != nil {
		return err
	}
	return os.WriteFile(destination, data, 0o666)
}

// BackupFiles creates <backups>/<YYYYMMDD>/<HHmmss>-<command>/<label> and
// copies all existing targets. Missing sources are reported in Skipped.
// Validation is completed before the first mkdir, so invalid labels leave no
// partial backup tree. The function returns an error for I/O and path errors.
func BackupFiles(paths core.HomerPaths, command string, targets []BackupTarget, options ...any) (BackupResult, error) {
	if err := assertSafeSegment(command, "command"); err != nil {
		return BackupResult{}, err
	}
	for _, target := range targets {
		if err := assertSafeSegment(target.Label, "label"); err != nil {
			return BackupResult{}, err
		}
	}
	modes, useModes, err := modeFromOptions(options)
	if err != nil {
		return BackupResult{}, err
	}

	now := time.Now()
	backupDir := filepath.Join(paths.BackupsDir, dateStamp(now), timeStamp(now)+"-"+command)
	if err := os.MkdirAll(backupDir, 0o777); err != nil {
		return BackupResult{}, err
	}

	result := BackupResult{BackupDir: backupDir, BackedUp: make([]string, 0, len(targets)), Skipped: make([]string, 0)}
	for _, target := range targets {
		if _, err := os.Stat(target.SourceAbs); err != nil {
			if os.IsNotExist(err) {
				result.Skipped = append(result.Skipped, target.Label)
				continue
			}
			return BackupResult{}, err
		}
		if err := copyPath(target.SourceAbs, filepath.Join(backupDir, filepath.FromSlash(target.Label))); err != nil {
			return BackupResult{}, err
		}
		result.BackedUp = append(result.BackedUp, target.Label)
	}

	if useModes {
		if modes.Dir != 0 {
			if err := chmodIfPresent(paths.BackupsDir, modes.Dir); err != nil {
				return BackupResult{}, err
			}
			if err := chmodIfPresent(filepath.Dir(backupDir), modes.Dir); err != nil {
				return BackupResult{}, err
			}
		}
		if err := applyModes(backupDir, modes); err != nil {
			return BackupResult{}, err
		}
	}
	return result, nil
}

// backupFiles is the package-local spelling used by migration-oriented tests.
func backupFiles(paths core.HomerPaths, command string, targets []BackupTarget, options ...any) (BackupResult, error) {
	return BackupFiles(paths, command, targets, options...)
}

// PruneResult contains date-directory names relative to BackupsDir.
type PruneResult struct {
	Removed []string
	Kept    []string
}

// PruneBackups retains the newest keep date directories. If keep is omitted,
// seven dates are retained. Non-date entries are intentionally untouched.
func PruneBackups(paths core.HomerPaths, keep ...int) (PruneResult, error) {
	retention := DefaultBackupKeep
	if len(keep) > 1 {
		return PruneResult{}, fmt.Errorf("pruneBackups: keep 只能传一个")
	}
	if len(keep) == 1 {
		retention = keep[0]
	}
	if retention < 0 {
		return PruneResult{}, fmt.Errorf("pruneBackups: keep 必须是非负整数（当前: %d）", retention)
	}

	entries, err := os.ReadDir(paths.BackupsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return PruneResult{Removed: []string{}, Kept: []string{}}, nil
		}
		return PruneResult{}, err
	}
	names := make([]string, 0)
	for _, entry := range entries {
		if entry.IsDir() && dateDirectoryPattern.MatchString(entry.Name()) {
			names = append(names, entry.Name())
		}
	}
	sort.Sort(sort.Reverse(sort.StringSlice(names)))
	kept := append([]string(nil), names[:min(retention, len(names))]...)
	removed := append([]string(nil), names[min(retention, len(names)):]...)
	for _, name := range removed {
		if err := os.RemoveAll(filepath.Join(paths.BackupsDir, name)); err != nil {
			return PruneResult{}, err
		}
	}
	return PruneResult{Removed: removed, Kept: kept}, nil
}

func pruneBackups(paths core.HomerPaths, keep ...int) (PruneResult, error) {
	return PruneBackups(paths, keep...)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
